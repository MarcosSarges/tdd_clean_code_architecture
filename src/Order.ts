import Cart from './Cart';
import Cupom from './Cupom';
import User from './User';

export default class Order {
  checkout(cart: Cart, user: User, cupom?: Cupom) {
    const paymentTotal = cupom
      ? cupom?.applyDiscount(cart)
      : cart.getCartDetails.totalPrice;

    return {
      user: user.userInfos,
      status: 'DONE',
      items: cart.getCartDetails.items,
      paymentTotal,
    };
  }
}
