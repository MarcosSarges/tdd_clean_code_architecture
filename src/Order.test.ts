// 1 - Não deve fazer um pedido com cpf inválido
// 2 - Deve fazer um pedido com 3 itens (com descrição, preço e quantidade)

import Order from './Order';

const ITEM = {
  description: 'Shot Azul',
  price: 1000,
  quantity: 1,
};

// 3 - Deve fazer um pedido com cupom de desconto (percentual sobre o total do pedido)
describe('Order', () => {
  let order: Order;

  beforeEach(() => {
    order = new Order();
  });

  it('Não deve fazer um pedido com cpf inválido', () => {
    const cpf = '553.566.310-55';
    order.setupUser(cpf);

    expect(() => {
      order.checkout();
    }).toThrowError('CPF INVALID');
  });
  it('Deve fazer um pedido com cpf válido', () => {
    const cpf = '553.566.310-73';
    order.setupUser(cpf);
    order.addItemInCart(ITEM);

    expect(order.checkout()).toMatchSnapshot();
  });

  it('Deve fazer um pedido com 3 itens (com descrição, preço e quantidade) ', () => {
    order.setupUser('553.566.310-73');

    order.addItemInCart(ITEM);
    order.addItemInCart(ITEM);
    order.addItemInCart(ITEM);

    expect(order.checkout()).toMatchSnapshot();
  });
  it('Deve fazer um pedido com cupom de desconto (percentual sobre o total do pedido)', () => {
    order.setupUser('553.566.310-73');
    order.addItemInCart(ITEM);
    order.addCupom('OFF50');

    expect(order.checkout().priceTotal).toEqual(500);
  });
  it('Não deve aplicar desconto caso o cupom seja invalido', () => {
    order.setupUser('553.566.310-73');
    order.addItemInCart(ITEM);
    order.addCupom('OFF10');

    expect(order.checkout().priceTotal).toEqual(1000);
  });
});
