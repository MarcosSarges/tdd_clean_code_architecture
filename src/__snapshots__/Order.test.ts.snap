// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Deve fazer um pedido com 3 itens (com descrição, preço e quantidade)  1`] = `
Object {
  "ordered": true,
  "priceTotal": 3000,
  "products": Array [
    Object {
      "description": "Shot Azul",
      "price": 1000,
      "quantity": 1,
    },
    Object {
      "description": "Shot Azul",
      "price": 1000,
      "quantity": 1,
    },
    Object {
      "description": "Shot Azul",
      "price": 1000,
      "quantity": 1,
    },
  ],
}
`;

exports[`Order Deve fazer um pedido com cpf válido 1`] = `
Object {
  "ordered": true,
  "priceTotal": 1000,
  "products": Array [
    Object {
      "description": "Shot Azul",
      "price": 1000,
      "quantity": 1,
    },
  ],
}
`;
